### ISSUE TITLE
`hostmetricsreceiver` semantic conventions transition

### ISSUE THREAD

**Issue by braydonk** [open-telemetry/opentelemetry-collector-contrib#35325]
### Component(s)

receiver/hostmetrics

### Is your feature request related to a problem? Please describe.

This issue will be used to track the work for making `hostmetricsreceiver` capable of transitioning to Semantic Conventions compliance.

### Describe the solution you'd like

Task list:

[] Add feature to `mdatagen` to allow it to generate packages with names other than `metadata`
[] Introduce a semantic conventions featuregate in the hostmetrics receiver
[] Add a semconv spec to each receiver and allow it to produce either existing or semconv metrics
  [] processscraper
  [] cpuscraper
  [] diskscraper
  [] filesystemscraper
  [] loadscraper
  [] memoryscraper
  [] networkscraper
  [] pagingscraper

**Comment by crobert-1** [comment]
_Removing `needs triage` as this makes sense to me, and was filed by a code owner._

**Comment by rogercoll** [comment]
>  Introduce a semantic conventions featuregate in the hostmetrics receiver

Do we plan to support the two metrics schemas during the transition? Or new naming features will be only applied to the semconv schema?

**Comment by braydonk** [comment]
> Do we plan to support the two metrics schemas during the transition? Or new naming features will be only applied to the semconv schema?

I think the plan at the moment is to support two schemas, but leaving it up to implementer whether a new metric or attribute change should be added to both schemas at the same time or only the semconv schema. Given that we are getting closer and closer to stabilization before I'm able to initiate this plan, I'm hoping we won't have to support two schemas for long. I think the two schema support will be necessary for a short while though, given that some of the metrics/attributes are a major lift that users need time to prepare for.

**Comment by ChrsMark** [comment]
Other areas like messaging SemConv define a specific migration plan for instrumentations, ie: https://github.com/open-telemetry/semantic-conventions/pull/1198/files#diff-bbe463329a83f81ca6c90d49d1998d69929432ba28bb6eed991a03cd110f74e6R12-R27

I'm preparing an early stage plan for k8s SemConv as well: https://github.com/open-telemetry/semantic-conventions/pull/1597.

What I'm missing though is what would be the plan to follow in the Collector.
1. Should that be an environment variable or a feature gate?
2. Should we support changing to only new SemConv but also new+old(dup)? See https://github.com/open-telemetry/semantic-conventions/pull/1597.
3. What versioning plan are we going to follow since the Collector has only one Major version so far and it's basically released on Minor version basis?

I would like to collect feedback here since answering the above questions will help us making our migration plans more concrete.

@open-telemetry/semconv-system-approvers @open-telemetry/collector-contrib-approvers could you please share your thoughts?

**Comment by mx-psi** [comment]
> Should that be an environment variable or a feature gate?

A feature gate is the idiomatic approach in the Collector, I think this is what our users would expect and what makes the most sense here.

> Should we support changing to only new SemConv but also new+old(dup)

Do we have data on what users preferred for this on the HTTP migration?

> What versioning plan are we going to follow since the Collector has only one Major version so far and it's basically released on Minor version basis?

Not sure I understand this question: what is the concern here? IMO doing this migration (at least until we have the beta stage) should be a requirement for the hosmetrics receiver or the k8s components to be marked as 1.0. Since we are at 0.x, breaking users is okay, but we want to do it slowly and allowing them to adopt this change at their own pace through the feature gate approach.

**Comment by braydonk** [comment]
> Should we support changing to only new SemConv but also new+old(dup)? See https://github.com/open-telemetry/semantic-conventions/pull/1597.

On first glance I struggle to see the usecase for hostmetrics to produce both schemas at the same time, but we can certainly find a way to build that in if users would be able to make use of it.

**Comment by mx-psi** [comment]
Discussed on 2024-11-21 meeting. We probably want two feature gates: one for enabling the new schema, and one for disabling the old schema. Then:
- On alpha old is enabled, new is disabled
- On beta/stable old is disabled, new is enabled
- It is an error to disable both

**Comment by ChrsMark** [comment]
@braydonk  is the first item in the description list covered or there is anything else missing in `mdatagen` for the transition plan?

For context, at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/40708#issuecomment-3027637883 it was raised how the transition plan will actually be performed, one PR VS multiple ones etc. So it'd worth it starting to make the plan more concrete.

/cc @evan-bradley

**Comment by braydonk** [comment]
Current status on `mdatagen` is that the feature to generate multiple packages is technically merged: https://github.com/open-telemetry/opentelemetry-collector/issues/11231

But there is a bug in that. I had a PR to fix it but that is stalled because there was disagreement with the idea that `mdatagen` generating multiple packages is something that should be done at all: https://github.com/open-telemetry/opentelemetry-collector/issues/11469

We are debating an alternative in: https://github.com/open-telemetry/opentelemetry-collector/issues/12620

I have an idea for how to do the transition in the multiple-package context. But with that idea still in contention, I can't really give any ideas for how exactly the transition would take place.

In the multiple package scenario, this was my hypothetical plan:

In the base receiver, two new featuregates would be introduced. One featuregate would represent the old schemas, and be on by default. The other would represent semconv schemas and be off by default. These overall featuregates would dictate the fallback behaviour for which schemas each scraper should produce.

Then for each scraper:

1. Pick a scraper to do the work for.
1. The scraper would be refactored to do all necessary stat collection up front rather than collecting at metric record time. This is to prepare for there being multiple packages. This would be its own PR for each scraper.
1. A new `semconv-metadata.yaml` would be added and a new package generated. This would be its own PR, and System Semconv WG members would be included to verify the metrics in the new schema match the semantic conventions.
1. The alternate path to generate whichever metric schema would be implemented. By this point the scraper should be collecting the metrics in the new order, which means that at recording time the featuregates would be checked to determine which schemas are produced (including the possibility that both are enabled). I am thinking there would be two featuregates at the scraper level, same setup as in the base receiver, which would take precedence just for that scraper. This would be its own PR.

The caveats for this plan are:

1. As I mentioned, the multiple package from `mdatagen` strategy is in contention.
2. This assumes that there isn't some kind of better solution, perhaps using Weaver, for generating Collector MetricBuilder stuff directly off of released Semantic Convention schemas. Realistically this would probably be better, but I built my plan off the assumption that I'm working with what we have now.

**Comment by ChrsMark** [comment]
Thank's @braydonk for the summary!

> In the base receiver, two new featuregates would be introduced.

For reference the K8s SIG will follow the same approach: https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-migration/

One thing that is still pending in System SemConv is writing a migration guide similar to the one we have already for K8s: https://github.com/open-telemetry/semantic-conventions/issues/1538

> This assumes that there isn't some kind of better solution, perhaps using Weaver, for generating Collector MetricBuilder stuff directly off of released Semantic Convention schemas. 

I think it would be super useful to be able to base code and docs generation based on released Semantic Conventions. I have filed an [issue](https://github.com/open-telemetry/opentelemetry-collector/issues/13297) for this which we can consider. My initial rough idea was to have `metadata.yaml` files to reference metrics from existing released Semantic Conventions, but I haven't really validated this idea. Do you think this could work here?

**Comment by braydonk** [comment]
Status Update:

We have settled on multiple package generation in `mdatagen` as the path forward for this.
 
The blocking bug on `mdatagen` was fixed, not by the linked PR but in the crossfire of a refactor. Lucky!

I have updated the task list to match the current intended set of work, with order considered.

I have opened an RFC for a major feature that will unblock unifying the `processesscraper` and `processscraper`: https://github.com/open-telemetry/opentelemetry-collector/pull/13358

**Comment by ChrsMark** [comment]
> Refactor each scraper to do collection in a way that separates it from reporting, so that the possibility of reporting values in different formats is easier to handle later

@braydonk another I issue I stumble upon is how we are going to handle the configuration. Will we merge both old and new metrics in the same section or we are going to separate them?

Specifically at https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1456a586ab7aebbc05a9f06c473f1bdcec99d685/receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go#L98 I assume we would need 2 different `MetricsBuilder`s, one for the `metadata` package and one for the `metadata_old`? However the https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/5ee95c6f8b886f24070b2938c3418d1cbbbeadd0/receiver/hostmetricsreceiver/internal/scraper/processscraper/config.go#L16 only defines the metric's part from one package. Should we have something like `metrics_legacy` and `metrics` in the config or I miss anything here?

**Comment by braydonk** [comment]
This was exactly the thing I stumbled into as well, which is why this first task was done:

> Add feature to mdatagen to allow it to generate packages with names other than metadata

There will be two metrics builders in the scraper, and one will come from the `metadata` package and one will come from another package generated from a separate `metadata.yaml` with a different name.

I think we will need two entries here, `metrics_legacy` and `metrics` sounds fine to me.

**Comment by ChrsMark** [comment]
> I think we will need two entries here, metrics_legacy and metrics sounds fine to me.

👍🏼 

Probably we will need some logic there so as to treat `metrics` as `metrics_legacy` if the latter is not defined so as to ensure current configs do not break 🤔 . Not sure if/how that's doable though.

**Comment by braydonk** [comment]
Ah yeah the forward compatibility it going to be a pain. I don't think we have aliases for config fields do we? Without that this might be very challenging to do without a major config breakage.

**Comment by ChrsMark** [comment]
Yeap, that's what I saw too. FWIWI, for internal telemetry that is not configurable it seems that the dual scheme works fine: https://github.com/open-telemetry/opentelemetry-collector/pull/14303

**Comment by yaten2302** [comment]
Hi @braydonk , can I help in this in some way 👀 
I read the above discussions & comments, I've got an outro of what this is all about, correct me if I'm wrong: The `hostmetricsreceiver` currently doesn't support semantic conventions. We need to migrate it to semconv in such a way that the users using this component shouldn't get affected. So, we can probably give them an option, to continue using the prev one or migrate to the new semconv.

However, I've a doubt here (pardon if it's a basic one, I'm currently trying to understand the internals so it's taking a bit time) - how can we get to know that a component is following the semconv or not? I noticed something in the [hostmetricsreceiver/cpuscraper](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/metadata.yaml#L11), it has `sem_conv_version: 1.9.0`, sem_conv version specified. However, this isn't there in other receivers, for say [redisreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/redisreceiver/metadata.yaml).

**Comment by braydonk** [comment]
I don't think this issue is ready for outside contribution yet. Might be once I've transitioned the first scraper to the new form, in which case other contributors can simply copy what I did. But there's some foundational work that still needs to get done before I can make that first scraper transition.

> I noticed something in the [hostmetricsreceiver/cpuscraper](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/metadata.yaml#L11), it has sem_conv_version: 1.9.0, sem_conv version specified.

This is unrelated to [System Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/system/README.md) that we are moving to. That semconv callout only exists for `mdatagen` templates, in particular [setting the `SchemaURL` for a resource](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b41687dc9c99a74552986c3d0698962452c83d5f/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go#L458) which exists despite us not following it as a requirement of the `mdatagen` templates.
